bundles/org.openhab.binding.philipsair/src/main/java/org/openhab/binding/philipsair/internal/connection/PhilipsAirCipher.java

```diff
@@ -0,0 +1,146 @@
+            16);
+    private static final Random RAND = new Random();
+
+    @Nullable
+    private Cipher decipher;
+    @Nullable
+    private Cipher cipher;
+    private final BigInteger a;
+    private final BigInteger aPow;
+
+    public PhilipsAirCipher() throws GeneralSecurityException {
+        // prepare numbers for key
+        a = randomForBitsNonZero(256, RAND);
+        aPow = G.modPow(a, P);
+    }
+
+    @SuppressWarnings("null")
+    public void initKey(String key) throws GeneralSecurityException {
+        try {
+            decipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+            decipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(HexUtils.hexToBytes(key), "AES"),
+                    new IvParameterSpec(new byte[16]));
+
+            cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(HexUtils.hexToBytes(key), "AES"),
+                    new IvParameterSpec(new byte[16]));
+        } catch (GeneralSecurityException e) {
+            logger.error("An exception occured", e);
+            cipher = null;
+            decipher = null;
+            throw e;
+        }
+    }
+
+    public String getApow() {
+        return aPow.toString(16);
+    }
+
+    private static BigInteger randomForBitsNonZero(int numBits, Random r) {
+        BigInteger candidate = new BigInteger(numBits, r);
+        while (candidate.equals(BigInteger.ZERO)) {
+            candidate = new BigInteger(numBits, r);
+        }
+        return candidate;
+    }
+
+    public String calculateKey(String hellman, String key) throws GeneralSecurityException, InterruptedException,
+            TimeoutException, ExecutionException, InvalidAlgorithmParameterException {
+        BigInteger b = new BigInteger(hellman, 16);
+        BigInteger s = b.modPow(a, P);
+        byte[] sByteArray = s.toByteArray();
+        // remove trailing 0
+        if (sByteArray.length > 128 && sByteArray[0] == 0) {
+            sByteArray = Arrays.copyOfRange(sByteArray, 1, 128);
+        }
+
+        byte[] sByteArrayTrunc = Arrays.copyOfRange(sByteArray, 0, 16);
+        byte[] hexKey = HexUtils.hexToBytes(key);
+
+        Cipher ciph = Cipher.getInstance("AES/CBC/PKCS5Padding");
+        ciph.init(Cipher.DECRYPT_MODE, new SecretKeySpec(sByteArrayTrunc, "AES"), new IvParameterSpec(new byte[16]));
+
+        byte[] keyDecoded = ciph.doFinal(hexKey);
+        String aesKey = HexUtils.bytesToHex(keyDecoded).substring(0, 32);
+        return aesKey;
+    }
+
+    public @Nullable String decrypt(String encodedContent) throws IllegalBlockSizeException, BadPaddingException {
+        if (decipher == null) {
+            return null;
+        }
+
+        @SuppressWarnings("null")
+        byte[] decoded = decipher.doFinal(Base64.getDecoder().decode(encodedContent));
+        byte[] unpaded = Arrays.copyOfRange(decoded, 2, decoded.length);
+        return new String(unpaded);

```

Comment:
The default charset varies from system to system so it is better to specify it yourself instead of using the default.

Reply:
Good point -  extracting to static field